Rework CONTRIBUTING file

tomschr opened the pull request:
This PR fixes #342 and contains:

* All prior content was moved to `RELEASE.rst`.
* In `CONTRIBUTING.rst`:
  * Encourage people to contribute :wink: 
  * Address target group
  * Use CONTRIBUTING.rst for contributors only
  * Provide step-by-step instructions which can be used as a checklist

----

This is just a draft. It's an example how such text could look like. Feel free to suggest improvements.

Some aspects are a bit more verbose; probably advanced developers don't need that. However, not every developer is the same, nor does everybody know about the Git workflow. :wink: 

It's up to you to decide if this is enough or it should be shortened.

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,138 @@
+Running the Test Suite
+----------------------
+
+We use the default unittest module and `tox`_ to run tests against all supported
```
we use `twisted.trial` to run the test.

tomschr replied:
Ahh right. My fault. Will correct that. :wink:

tomschr replied:
Fixed in 601ea63.

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,138 @@
+
+You can decide to run the complete test suite or only part of it:
+
+* To run all tests, use::
```
this looks like a tox tutorial... we can keep it

With this the CONTRIBUTING file is more verbose... and more to read or skip for people who already know tox.

My opinion :)

tomschr replied:
Absolutely valid. :+1: It was just an example. 

If you think, your user base do know all about tox and how to use it, then we could leave that out. I have absolutely no problem to remove it or shorten it. :wink: 

Probably we can shorten it and provide a link for further details.

tomschr replied:
OK, revised the "Running the Test Suite" section and shortened it (commit 436795a).

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,138 @@
+
+      $ tox -av
+
+* To run only a specific test only, use the ``towncrier.test.FILE.CLASS.METHOD`` syntax.
```
for TDD I create a separate virtual env and call `trial` directy, include `trial -b towncrier` to drop me into a PDB session on errors so that I  can investigate what is wrong.

We can keep it. From my experience, tox is super slow of TDD.

tomschr replied:
Ahh, now it gets interesting! :slightly_smiling_face:

I'm actually not really sure if we should remove that all and replace it with just a link to tox. This goes into kind of recommendations and how to develop things in towncrier.

Should we add that here or somewhere else?

I think, a general direction would be nice as this helps new contributors.

tomschr replied:
OK, I've revised the "Running the Test Suite" section and shortened it in commit 436795a.

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,138 @@
+      $ tox -e py36-tests,py37-tests -- \
+        towncrier.test.test_project.InvocationTests.test_version
+
+Our code is checked against formatting issues with `flake8`_ issues.
```
using tox for flake8 is ok... but there are multiple check.

I have no idea why I can't suggest changes here.

A pre-commit run should look like this.

```
tox -e flake8,check-manifest,check-newsfragment
```

tomschr replied:
Good idea! This is also integrated in commit 436795a.

adiroiban reviewed (approved):
Looks good. Only minor changes.

I think the main issue with the current version is that it talks about stdlib unittest and we use twisted.trial 

In terms of contributing, maybe as part of this PR we can also add a PR template in which we add a quick checklist for new contributors

* Make sure changes are covered by existing or new tests
* Make sure local test run is green for at least one python version
* Create a file in src/towncrier/newsfragments/
* Make sure all CI tests are green  (this is checking all of the above)

--------------

My comments about tox are optional.

I think that the current PR is already an improvement over what we have in trunk so it can be merged.

But please consider mentioning that we use `trial` and maybe also run not only flake8 but also `tox -e flake8,check-manifest,check-newsfragment`

Thanks!

tomschr commented:
> I think the main issue with the current version is that it talks about stdlib unittest and we use twisted.trial [...]
> But please consider mentioning that we use trial and maybe also run not only flake8 but also tox -e flake8,check-manifest,check-newsfragment

Sure! I will change that. 

Regarding tox, I was thinking about opening another issue about tox itself. IMHO, the `tox.ini` could be improved to add a `description` key for each target. This allows tox to be called as `tox -av` and outputs a nice summary of all the targets. But it's only tangentially related to this issue. :wink: 

> In terms of contributing, maybe as part of this PR we can also add a PR template in which we add a quick checklist for new contributors

Great idea! :+1: I will add that too.

> I think that the current PR is already an improvement over what we have in trunk so it can be merged.

Thanks. :blush: Let me correct the your above issues first. :wink:

tomschr commented:
> In terms of contributing, maybe as part of this PR we can also add a PR template

I've created a first draft in commit 0de1289. Not sure if this is enough. Probably something is missing.

Let me know what you think about.

adiroiban reviewed (changes requested):
Thanks. Looks good...

But as mentioned over IRC there are broken links

I am using this page to view the rendered version https://github.com/tomschr/towncrier/blob/feature/342-contributing/CONTRIBUTING.rst

tomschr commented:
> [...] there are broken links

You mean probably the lines which contain `:ref:`? Well, I've recognized that too. However, it's a valid syntax construct in RST/Sphinx. Unfortunately, it isn't rendered correctly in GitHub.

For the time being, I could replace them. However, if you plan a documentation which goes beyond little READMEs, I would highly recommend to use them.

For example, if you think a short user guide could be helpful, we could host the HTML version on ReadTheDocs and link to the rendered version. This would avoid these "broken" links. :wink:

adiroiban commented:
The main broken link is 

> Open a new topic on our **GitHub discussion page**.

But I think that is also best to don't have RST references so that the page looks ok with GitHub rendering.

Thanks

tomschr commented:
> The main broken link is
> 
> > Open a new topic on our **GitHub discussion page**.
> 
> But I think that is also best to don't have RST references so that the page looks ok with GitHub rendering.

Not sure I understand you correctly. Maybe your definition of "broken" is different than mine. :wink: Can you explain why do see the link as "broken"?
It's the same with the issue tracker, only the texts are different. Both are rendered just fine in GitHub: They are texts that are linked to a URL.

The main idea was most of our users will read this text online. They just have to click the text to get redirected. IMHO it's not so important to actually see the link, it's more important that the link is correct.

Or do you prefer the link shown literally? Like in this step:

    Fork our project on GitHub using this link: https://github.com/twisted/towncrier/fork

It's possible, but it makes the text longer.

tomschr commented:
> But I think that is also best to don't have RST references so that the page looks ok with GitHub rendering.

I've updated the file now, find the latest [rendered output in my fork](https://github.com/tomschr/towncrier/blob/feature/342-contributing/CONTRIBUTING.rst).  :slightly_smiling_face: 

Better?

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -1,86 +1,127 @@
-    Commands are written with Linux in mind and a ``venv`` located in ``venv/``.
-    Adjust per your OS and virtual environment location.
-    For example, on Windows with an environment in the directory ``myenv/`` the Python command would be ``myenv/scripts/python``.
+* Open a new topic on our `GitHub discussion <gh_discussions>`_ page.
```
This is rendered as 
```
<p>Open a new topic on our <a href="/tomschr/towncrier/blob/feature/342-contributing/gh_discussions">GitHub discussion</a> page.</p>
```

```suggestion
* Open a new topic on our  `GitHub Discussions`_ page.
```

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -1,86 +1,127 @@
+.. ### Links
+
+.. _flake8: https://flake8.rtfd.io
+.. _gh_discussions: https://github.com/twisted/towncrier/discussions
```
```suggestion
.. _GitHub Discussions: https://github.com/twisted/towncrier/discussions
```

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -1,86 +1,127 @@
+  Or you would like to share your ideas with us.
 
-- Create a release branch with a name of the form ``release-19.9.0`` starting from the ``master`` branch.
+* Open a new `issue <issues>`_.
```
I don't think that this syntax works

Maybe something like this:

```suggestion
* Open a new issue using GitHub `issues`_ trackerr.
```

adiroiban commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,10 @@
+* [ ] Make sure changes are covered by existing or new tests.
+* [ ] Make sure local test run is green for at least one Python version.
+* [ ] Create a file in `src/towncrier/newsfragments/` for our changelog.
+* [ ] Make sure all CI tests are green (this is checking all of the above)
```
```suggestion
* [ ] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
```

adiroiban commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,10 @@
+<!-- add a "X" inside the brackets to confirm the point -->
+* [ ] Make sure changes are covered by existing or new tests.
+* [ ] Make sure local test run is green for at least one Python version.
+* [ ] Create a file in `src/towncrier/newsfragments/` for our changelog.
```
For towncrier we have `NEWS`.

I tried to inform users that the content of the file should not be a commit message, but it should contain important public information for other users of the project.

Feel free to repharse

```suggestion
* [ ] Create a file in `src/towncrier/newsfragments/` describing your change and that will be included in the public release notes.
```

adiroiban commented on CONTRIBUTING.rst:
```diff
@@ -1,86 +1,127 @@
+   c. Document any user facing changes in the ``README.rst`` file.
 
-- If the final release has been completed, continue below.
+   d. Create a newsfragment in ``src/towncrier/newsfragments/``.
```
```suggestion
   d. Create a newsfragment in ``src/towncrier/newsfragments/`` describing the changes and containing information that are of interest for end-users.
```

adiroiban reviewed (commented):
Sorry for being brief.
I tried to comment inline on broken links.

Maybe also add info about the Freenode #twisted-dev IRC channel. Some towncrier devs hang out there.

Thanks!

adiroiban commented:
> If yes, I can squash the commits into some logical units.

Thanks. Yes. You can do that before a merge.

altendky commented:
I didn't dig into the details on this but it does seem like you can have internal links in rst on GitHub.

https://gist.github.com/ionelmc/e876b73e2001acd2140f#211inline-markup

I'll note that several 'regular' links were also broken for me.  Both broken in terms of linking to the incorrect place (not valid pages) and in terms of rendering.  Anyways, I'll take a look again after the above comments are handled.

tomschr commented:
> Maybe also add info about the Freenode #twisted-dev IRC channel.

Good idea! :+1:  I've included it in the file now.

>  squash the commits into some logical units.

Done. :slightly_smiling_face: 

@altendky Kyle, that would be great if you could have a look too! :+1:

altendky reviewed (commented):
@tomschr, thanks for all the work here.  I know it takes awhile to write up documentation like this, hence us not having it...  until now.  `:]`  I've got a few more suggestions and corrections.  I don't like to use other people as my keyboard, but I also don't like to just take over their PRs.  I'm game for addressing my own concerns here if you prefer.

Thanks for your help here and elsewhere recently.

tomschr commented on CONTRIBUTING.rst:
```diff
@@ -1,86 +1,127 @@
-- If another release candidate is required:
 
-  - Submit PRs against the release branch to integrate the needed changes.  Any PRs could be cherry picks from the ``master`` branch if already resolved there, or direct PRs against the release branch that will be merged back into master at the completion of the release.
+#. In your local repository create a pull request:
```
Well, the basic idea was, that the user creates a pull request in the end. What is needed is described in the substeps of this.
Maybe the first part is misleading. We could rephrase that and write: "Prepare a pull request:".

altendky replied:
Your suggestion is good by me too.

```suggestion
#. Prepare a pull request:
```

My nitpick here was that the branch is created in your local repository (and pushed to your fork repository on GitHub) while the pull request itself is created in the 'official' repository on GitHub.

tomschr replied:
You are absolutely right. Thanks for your comment! :+1:

tomschr commented:
Thanks @altendky, your feedback is very appreciated! :+1: 

> @tomschr, thanks for all the work here. I know it takes awhile to write up documentation like this, hence us not having it... until now. `:]` 

Happy to help here. Actually I saw, well, the need when I first made my contribution. :wink: 

> I've got a few more suggestions and corrections. I don't like to use other people as my keyboard, but I also don't like to just take over their PRs. I'm game for addressing my own concerns here if you prefer.

That's perfectly fine! :+1: Thank you and don't worry. I expected that this PR creates some discussion and work. That is natural when you write documentation. It's my daily job, so writing is rewriting. :wink:

You found some important issues which I'm very happy you addressed. I already included it. There is only one item in your list that I would like to discuss with you. Maybe I misunderstood your idea.


> Thanks for your help here and elsewhere recently.

No problem. :slightly_smiling_face:

tomschr commented:
I've squashed it and added both of you as co-authors.

After looking over it again, the "Ways to communicate and contribute" section could be split into two separate sections: "Ways to communicate" and "Ways to contribute".

The idea is to put into the first section the GitHub discussion and the item about IRC. The section about contribution contains the rest.

Would that make sense?

altendky commented:
I'm fine with it split or not.  Whatever you prefer.  I think that I only noticed we had discussions when I was reading through this.  I know some people are adamant that discussions can't happen in issues...  I'm not so strict and don't know how much value discussions bring vs. issues with a 'question' tag or whatever.  But, that's a separate topic from this PR.

tomschr commented:
Thanks @altendky for your answer. :+1: I'm fine to leave it as is. If you like the current state, you can merge it. Let me know if you have further changes.

Thanks!

altendky commented:
@adiroiban, just checking back to see if your requested change was addressed.

adiroiban commented:
So.... after #368 is merged (since #89) is stalled we can integrate this docs into the main doc.

adiroiban reviewed (approved):
Let's merge this as this is already a big step.

I did a small update to include info about pre-commit.

I hope we can include check-manifest and check-towncrier as part of the standard pre-commit framework
